ion-schema/src/main/kotlin/com/amazon/ionschema/model/DiscreteRange.kt

```diff
@@ -0,0 +1,41 @@
+package com.amazon.ionschema.model
+
+/**
+ * A range over a type that is a finite or countably infinite set. The values contained in the range include both of the
+ * limits—[start] and [endInclusive].
+ *
+ * A `DiscreteRange` is a _bounded, closed interval_ when both limits are non-null, and it is a _half-bounded interval_
+ * when one of the limits is `null`. At least one of [start] and [endInclusive] must be non-null.
+ *
+ * A `DiscreteRange` is allowed to be a _degenerate interval_ (i.e. `start == endInclusive`), but it may not be an
+ * empty interval (i.e. `start > endInclusive`).
+ */
+data class DiscreteRange<T : Comparable<T>>(val start: T?, val endInclusive: T?) {
```

Comment:
Discrete Ranges don't require us to store whether the boundary value is exclusive because we can "translate" exclusive to +/-1 of the boundary value.

More importantly, however, is the fact that the logic for determining emptiness is different. With a discrete range, it is true that `∀n∈Z, ¬∃x∈Z, n<x & x<n+1`, but for a range over continuous values, it would not be empty. If `n` and `x` are both in the set of all rational numbers rather than the set of all integers, then we can trivially prove that there does exist at least one `x`, such as `x = n + 0.5`.

Finally, my admittedly subjective opinion is that it helps the reader/user understand how the ranges are to be used by keeping them distinct.

Reply:
But whether a range is continuous or discrete depends on the type of arguments, right? I don't see any constraints that model that... so it looks like I can define a Discrete range of a Continuous (but Comparable) type and vice versa.

Further I don't see any definition in the spec about emptiness or degenerate-ness so I don't see where the requirement for the validation (and the need to split it) even comes from.

What I do see is what _appears to be_ duplicated code. 🤷

Reply:
The test suite requires that ranges (at least for countable things like number of codepoints) are non-empty, and I believe that requirement has been part of the reference implementation since the initial release. https://github.com/popematt/ion-schema-tests/blob/master/ion_schema_2_0/constraints/codepoint_length.isl#L79

At this point, the non-empty range requirement is _de facto_ part of the specification, and the specification document should be updated to reflect that.

Degenerate ranges are fine. I could remove that bit of the sentence if it's too confusing to have it, though.

Reply:
I have found a clean way to avoid the duplicated code, though, so I will remove the `DiscreteRange<T>` implementation.